### PR TITLE
Add support to optionally allow surrogate pair entities (#165)

### PR DESCRIPTION
**Pull request by Magmaruss**
This feature is an answer for the issue #165. It's extending the XML reader functionality with a new configuration option (default disabled) to support reading XML with unicode characters written using surrogate pair (UTF-16 encoding). This feature is very important to correctly read XML from the legacy services which produces responses this way and there is no way to change the behavior of them.

Example:
The external service produces a response with a unicode character written using UCS2:
```
<response>
    <value>Merry Christmas &#55356;&#57221;</value>
</response>
```
Default behavior of XMLStreamReader is throwing an exception with message:
```
Illegal character entity: expansion character (code 0xd83c
```
After setting the `P_RESOLVE_ENTITY_SURROGATE_PAIRS` parameter to `true` it will be readed with no exception converting the surrogate pair entities to the appropriate unicode (🎅)

### CODE REVIEW

**Comment by cowtowncoder** [comment]
First of all: thank you for contributing this feature.

Second: could you add a description that explains why this is needed, what is the use case (ideally with a simple example) and so on?

**Comment by Magmaruss** [comment]
Description added.

File: src/main/java/com/ctc/wstx/api/ReaderConfig.java [hunk 1]
```diff
@@ -418,6 +422,11 @@ public final class ReaderConfig
      * references
      */
     protected URL mBaseURL;
+    
+    /**
+     * Resolve surrogate pairs of entities (2 code-points as one target character)
+     */
+    protected boolean mResolveEntitySurrogatePairs = Boolean.FALSE;
```
Comment (gnodet) [inline src/main/java/com/ctc/wstx/api/ReaderConfig.java:429]: `false` ?

Comment (Magmaruss) [reply src/main/java/com/ctc/wstx/api/ReaderConfig.java:429]: Due to task specification from the issue #165. Quotation:

> If you have time and interest, adding new configuration property that would allow inclusion of surrogates could be acceptable as well: **something disabled by default, but that can be enabled**.

File: src/main/java/com/ctc/wstx/api/WstxInputProperties.java [hunk 2]
```diff
@@ -300,6 +300,13 @@ public final class WstxInputProperties
      * DTD subset).
      */
     public final static String P_BASE_URL = "com.ctc.wstx.baseURL";
+    
+    /**
+     * Property of type {@link java.lang.Boolean}, that will allow parsing
+     * high unicode characters written by surrogate pairs (2 code points)
+     * Default set as Boolean.FALSE, because it is not a standard behavior
+     */
```
Comment (cowtowncoder) [inline src/main/java/com/ctc/wstx/api/WstxInputProperties.java:308]: Let's add

    @since 6.6

(next version due to additional functionality and not just bug fixes)

File: src/main/java/com/ctc/wstx/sr/StreamScanner.java [hunk 3]
```diff
@@ -1188,55 +1188,121 @@ protected int resolveSimpleEntity(boolean checkStd)
...
+            
             /* We get here either if we got it all, OR if we ran out of
              * input in current buffer.
              */
             if (c == ';') { // got the full thing
                 mInputPtr = ptr;
-                validateChar(value);
+                
+                if (mConfig.isResolvingEntitySurrogatePairs() && pairValue > 0) {
+                    /*
```
Comment (cowtowncoder) [inline src/main/java/com/ctc/wstx/sr/StreamScanner.java:1294]: add reference to `[woodstox-core#165]`

File: src/test/java/org/codehaus/stax/test/stream/TestEntityRead.java [hunk 4]
```diff
@@ -407,6 +457,35 @@ public void testInvalidEntityPartial()
...
+            "<root>surrogate pair: &#55356;&#553</root>"
+        };
+        
+        for (String invalidSurrogatePair: invalidSurrogatePairs) {
+            XMLStreamReader sr = getReader(invalidSurrogatePair,
+                    true, false, true, true);
+            try {
+                streamThrough(sr);
+                fail("Expected an exception for invalid surrogate pair");
+            } catch (Exception e) { }
```
Comment (cowtowncoder) [inline src/test/java/org/codehaus/stax/test/stream/TestEntityRead.java:486]: Would be better to catch specific `Exception` subtype and check exception message that it fails for expected reason (just match a phrase, no need to match exact message).

**Comment by cowtowncoder** [comment]
First of all: apologies for ultra-slow follow up here. But I am ready to get this merged.

One minor legal thing first: if we haven't yet gotten a CLA, we'd need one from here (it's under Jackson project, that's fine):

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

and it's easiest to print, fill & sign, scan/photo, email to `cla` at fasterxml dot com.
(for project you can add Woodstox or FasterXML/Woodstox).

This is only needed once for all contributions to Woodstox, and once I get it, I can proceed with merge.
I added some notes already; just need to re-read main logic and should be good to merge.

Thank you again for this contribution!

File: src/main/java/com/ctc/wstx/sr/StreamScanner.java [hunk 5]
```diff
@@ -1188,55 +1188,121 @@ protected int resolveSimpleEntity(boolean checkStd)
...
+                     * If pair value is not in range of low surrogate values, then throw an error
+                     */
+                    if (pairValue < 0xDC00 || pairValue > 0xDFFF) {
+                        reportNoSurrogatePair(value);
```
Comment (cowtowncoder) [inline src/main/java/com/ctc/wstx/sr/StreamScanner.java:1298]: Would be better if we reported both leading surrogate and second non-surrogate entity value?

File: src/main/java/com/ctc/wstx/sr/StreamScanner.java [hunk 6]
```diff
@@ -1188,55 +1188,121 @@ protected int resolveSimpleEntity(boolean checkStd)
...
-                    if (c <= '9' && c >= '0') {
-                        value += (c - '0');
-                    } else if (c >= 'a' && c <= 'f') {
-                        value += (10 + (c - 'a'));
-                    } else if (c >= 'A' && c <= 'F') {
-                        value += (10 + (c - 'A'));
-                    } else {
-                        mInputPtr = ptr; // so error points to correct char
-                        throwUnexpectedChar(c, "; expected a hex digit (0-9a-fA-F).");
+            boolean isValueHighSurrogate = false;
```
Comment (cowtowncoder) [inline src/main/java/com/ctc/wstx/sr/StreamScanner.java:1194]: Ok, so, I assume this works, but is quite hard to follow. What I'd suggest is to split code in such a way that the original decoding for the first entity is within initial method; and the rest of processing in separate method, only called if:

1. Entity surrogate pair decoding is enabled
2. First entity is valid first surrogate

This probably leads to slight duplication of code (decoding loop) but allows removing the outer loop since each method only decodes one entity.
And if duplicate looks too ugly decoding loops can be extracted into separate method too.

File: src/main/java/com/ctc/wstx/api/ReaderConfig.java (hunk 1 above)
Comment (gnodet) [reply src/main/java/com/ctc/wstx/api/ReaderConfig.java:429]: Ah, sure.  My point was that, given the property is the scalar type `boolean`, the default value could `false` instead of `Boolean.FALSE`.  No big deal though...

Comment (cowtowncoder) [reply src/main/java/com/ctc/wstx/api/ReaderConfig.java:429]: ahhhh. Yes, that makes sense to me. It should be `false` indeed.

**Comment by cowtowncoder** [comment]
@Magmaruss quick note: I gave a wrong email address to send CLA to -- CLA actually needs to be sent to `info` at fasterxml dot com instead; cla email alias did not yet exist unlike I thought.
So just in case you already sent one, could you please re-send it. Apologies for mix-up.

**Comment by cowtowncoder** [comment]
NOTE: CLA received; can now focus on finalizing PR.

File: src/main/java/com/ctc/wstx/api/WstxInputProperties.java [hunk 7]
```diff
@@ -300,6 +300,13 @@ public final class WstxInputProperties
...
+    public final static String P_RESOLVE_ENTITY_SURROGATE_PAIRS = "com.ctc.wstx.resolveEntitySurrogatePairs";
```
Comment (cowtowncoder) [inline src/main/java/com/ctc/wstx/api/WstxInputProperties.java:309]: Will rename to `P_ALLOW_SURROGATE_PAIR_ENTITIES` (same for constant).

**Comment by cowtowncoder** [comment]
@Magmaruss I hope you don't mind my applying changes I suggested here, wrt naming and adding `@since` tags. I am open to different naming too if you strongly prefer yet different naming.

Beyond this, I think it'd be good to refactor decoding loop, possibly the way I suggested (but not necessarily, there may well be other ways to improve readability).

Also thank you for getting CLA; I am hoping we can get this PR merged soon!

File: src/main/java/com/ctc/wstx/sr/StreamScanner.java (hunk 6 above)
Comment (Magmaruss) [reply src/main/java/com/ctc/wstx/sr/StreamScanner.java:1194]: I'll try to refactor this part in my free time (maybe on the weekend), but my first assumptions was to write functional code with minimal changes in the class architecture, to not conflict your rules and coding style. I haven't seen before any contributor guide, so I wanted to be careful and do this most safety way. Please give me a time, because I don't have much free time after hours

Comment (cowtowncoder) [reply src/main/java/com/ctc/wstx/sr/StreamScanner.java:1194]: @Magmaruss understood! I could try my hand at change, but was worried I might get some parts wrong easily.
And my suggestion was not meant as critique; I much appreciate the contribution and it follows all reasonable expectations (esp. in absence of much documentation on what is expected of contributions).

File: src/main/java/com/ctc/wstx/sr/StreamScanner.java [hunk 8]
```diff
@@ -1186,57 +1186,69 @@ protected int resolveSimpleEntity(boolean checkStd)
 
         // Numeric reference?
         if (c == '#') {
-            c = buf[ptr++];
             int value = 0;
+            int pairValue = 0;
             int inputLen = mInputEnd;
-            if (c == 'x') { // hex
-                while (ptr < inputLen) {
+            final StringBuffer buffer = new StringBuffer(new String(buf));
```
Comment (cowtowncoder) [inline src/main/java/com/ctc/wstx/sr/StreamScanner.java:1192]: Ouch. This is not great; it'll create 8kB String just to parser a short entity.

File: src/test/java/org/codehaus/stax/test/stream/TestEntityRead.java [hunk 9]
```diff
@@ -73,6 +77,52 @@ public void testValidCharEntities()
         assertTokenType(END_ELEMENT, type);
         sr.close();
     }
+    
+    /**
+     * This unit test checks that resolving of surrogate pairs works
+     * as expected, including different ways of writing entities
+     */
+    public void testValidSurrogatePairEntities()
```
Comment (cowtowncoder) [inline src/test/java/org/codehaus/stax/test/stream/TestEntityRead.java:85]: Shouldn't there be a test or two for invalid ones too?

File: src/main/java/com/ctc/wstx/sr/StreamScanner.java (hunk 8 above)
Comment (cowtowncoder) [reply src/main/java/com/ctc/wstx/sr/StreamScanner.java]: Actual n/m, I was able to remote it without ill effects it seems (was not being used -- unless it's something untested).

File: src/test/java/org/codehaus/stax/test/stream/TestEntityRead.java (hunk 9 above)
Comment (Magmaruss) [reply src/test/java/org/codehaus/stax/test/stream/TestEntityRead.java:85]: Test for invalid ones is already there - `testInvalidSurrogatePairEntities` method: https://github.com/FasterXML/woodstox/pull/174/files/eb5189d107d911709774c79e1ed445865567d7ff#diff-50c2abc037511dda94147cf808bde7666e630ad76d08151e92ff86fb56b6f17aR466

**Comment by Magmaruss** [comment]
I can write simple test for disabled option, but for now when disabled it's throwing "Illegal character entity..." - default behavior before introducing the feature. Should I replace this exception with something special when first valid high surrogate detected or leave the default behavior and add test only?

**Comment by cowtowncoder** [comment]
> I can write simple test for disabled option, but for now when disabled it's throwing "Illegal character entity..." - default behavior before introducing the feature. Should I replace this exception with something special when first valid high surrogate detected or leave the default behavior and add test only?

Either way works for me: obviously improved error messages are always a plus, but if you want to limit the scope of changes & extra work, verifying default behavior is fine also.

File: src/test/java/org/codehaus/stax/test/stream/TestEntityRead.java (hunk 9 above)
Comment (cowtowncoder) [reply src/test/java/org/codehaus/stax/test/stream/TestEntityRead.java:85]: Right, I meant the case when new option was disabled, as per your other comment.

**Comment by cowtowncoder** [comment]
Hi @Magmaruss! I was wondering if you might be able to add basic unit test? No rush if you don't have time -- I just hope to be able to merge this in, get 6.6.0 released, and then be able to do #134 (increase min JDK baseline to 8 for Woodstox).

**Comment by Magmaruss** [comment]
Yes, sorry for that, but i am having a very difficult time. I will try to deliver it tonight or tomorrow at the latest

**Comment by cowtowncoder** [comment]
@Magmaruss Sorry did not mean to add pressure at all. I appreciate your help here and please take your time -- I don't want this to be another stress point.

**Comment by Magmaruss** [comment]
Test case for disabled option added. Please tell if you need something more.

**Review by cowtowncoder — APPROVED** [review]
Ok good, let's merge this!
